File: 1675-minimize-deviation-in-array/1675-minimize-deviation-in-array.cpp
class Solution {
public:
    int minimumDeviation(vector<int>& nums) {
        multiset<int,greater<int>>st;        
        for(int i=0;i<nums.size();i++)
        {
            if(nums[i]&1)
                st.insert(nums[i]*2);
            
            else
                st.insert(nums[i]);                
        }
        
        int ans = *(st.begin()) - *(st.rbegin());
        
        while(*st.begin()%2==0)
        {
            int val=*st.begin();
            st.erase(val);
            st.insert(val/2);
            
            ans= min(ans, (*(st.begin()) - *(st.rbegin())));
        }
        
        return ans;
        
        
        
        
        
        
    }
};

File: 1675-minimize-deviation-in-array/1675-minimize-deviation-in-array.java
import java.util.*;

class Solution {
    public int minimumDeviation(int[] nums) {
        TreeSet<Integer> treeSet = new TreeSet<>();
        int mini = Integer.MAX_VALUE,ans = Integer.MAX_VALUE;

      
        for (int num : nums) {
            int val = (num % 2 == 0) ? num : num * 2;
            treeSet.add(val);
            mini = Math.min(mini, val);
        }

        while (true) {
            int maxVal = treeSet.last(); 
            ans = Math.min(ans, maxVal - mini);

            if (maxVal % 2==1) {
                break;
            }
            
            treeSet.remove(maxVal);
            int halvedVal = maxVal / 2;
            treeSet.add(halvedVal);
            mini = Math.min(mini, halvedVal);
        }

        return ans;
    }
}
